rustls/src/server/hs.rs

```diff
@@ -239,52 +232,72 @@ impl ExtensionProcessing {
         config: &ServerConfig,
         cx: &mut ServerContext<'_>,
     ) -> Result<(), Error> {
-        let requires_client_rpk = config
-            .verifier
-            .requires_raw_public_keys();
-        let client_offers_rpk = hello
+        let client_supports = hello
             .client_certificate_extension()
-            .map(|certificate_types| certificate_types.contains(&CertificateType::RawPublicKey))
-            .unwrap_or(false);
+            .map(|certificate_types| certificate_types.to_vec())
+            .unwrap_or_default();
 
-        let raw_key_negotation_params = RawKeyNegotiationParams {
-            peer_supports_raw_key: client_offers_rpk,
-            local_expects_raw_key: requires_client_rpk,
-            extension_type: ExtensionType::ClientCertificateType,
-        };
         self.process_cert_type_extension(
-            raw_key_negotation_params.validate_raw_key_negotiation(),
+            client_supports,
+            config
+                .verifier
+                .requires_raw_public_keys(),
+            ExtensionType::ClientCertificateType,
             cx,
         )
     }
 
     fn process_cert_type_extension(
         &mut self,
-        raw_key_negotiation_result: RawKeyNegotationResult,
+        client_supports: Vec<CertificateType>,
+        requires_raw_keys: bool,
+        extension_type: ExtensionType,
         cx: &mut ServerContext<'_>,
     ) -> Result<(), Error> {
-        match raw_key_negotiation_result {
-            RawKeyNegotationResult::Negotiated(ExtensionType::ClientCertificateType) => {
+        debug_assert!(
+            extension_type == ExtensionType::ClientCertificateType
+                || extension_type == ExtensionType::ServerCertificateType
+        );
+        let raw_key_negotation_result = match (
+            requires_raw_keys,
+            client_supports.contains(&CertificateType::RawPublicKey),
+            client_supports.contains(&CertificateType::X509),
+        ) {
+            (true, true, _) => Ok((extension_type, CertificateType::RawPublicKey)),
+            // XXX XXX TRY REPLACING THIS WITH FINER CASES BELOW
+            // (false, _, true) => Ok((extension_type, CertificateType::X509)),
+            // XXX THIS CASE SEEMS TO BE NEVER TESTED - NOT IN MAIN CRATE TESTS & NOT IN `openssl-tests`
+            (false, false, true) => panic!("XXX"),
+            // XXX THIS CASE ONLY TRIGGERS FAILURE IN `openssl-tests` - DOES NOT TRIGGER FAILURE IN MAIN CRATE
+            // (false, true, true) => panic!("XXX"),
+            // XXX FINISH WITH NO CERT TYPE EXTENSION IN THIS CASE - DOES NOT SEEM TO TRIGGER ANY TEST FAILURES
+            (false, true, true) => return Ok(()),
```

Comment:
Considering that these updates which I have injected do not trigger any testing or CI failures, I think this DOES indicate a need for some better testing of these cases. I hope to investigate & contribute testing improvements to upstream RUSTLS in the near future (apologies no promise).